flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcFileWriter.java

```diff
@@ -0,0 +1,269 @@
+import org.apache.orc.CompressionKind;
+import org.apache.orc.OrcFile;
+import org.apache.orc.TypeDescription;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.orc.OrcBatchReader.schemaToTypeInfo;
+
+/**
+ * A {@link Writer} that writes the bucket files as Hadoop {@link OrcFile}.
+ *
+ * @param <T> The type of the elements that are being written by the sink.
+ */
+public class OrcFileWriter<T extends Row> extends StreamWriterBase<T> {
+
+	private static final long serialVersionUID = 3L;
+
+	/**
+	 * The description of the types in an ORC file.
+	 */
+	private TypeDescription schema;
+
+	/**
+	 * The schema of an ORC file.
+	 */
+	private String metaSchema;
+
+	/**
+	 * A row batch that will be written to the ORC file.
+	 */
+	private VectorizedRowBatch rowBatch;
+
+	/**
+	 * The writer that fill the records into the batch.
+	 */
+	private OrcBatchWriter orcBatchWriter;
+
+	private transient org.apache.orc.Writer writer;
+
+	private CompressionKind compressionKind;
+
+	/**
+	 * The number of rows that currently being written.
+	 */
+	private long writedRowSize;
+
+	/**
+	 * Creates a new {@code OrcFileWriter} that writes orc files without compression.
+	 *
+	 * @param metaSchema The orc schema.
+	 */
+	public OrcFileWriter(String metaSchema) {
+		this(metaSchema, CompressionKind.NONE);
+	}
+
+	/**
+	 * Create a new {@code OrcFileWriter} that writes orc file with the gaven
+	 * schema and compression kind.
+	 *
+	 * @param metaSchema      The schema of an orc file.
+	 * @param compressionKind The compression kind to use.
+	 */
+	public OrcFileWriter(String metaSchema, CompressionKind compressionKind) {
+		this.metaSchema = metaSchema;
+		this.schema = TypeDescription.fromString(metaSchema);
+		this.compressionKind = compressionKind;
+	}
+
+	@Override
+	public void open(FileSystem fs, Path path) throws IOException {
+		writer = OrcFile.createWriter(path, OrcFile.writerOptions(fs.getConf()).setSchema(schema).compress(compressionKind));
+		rowBatch = schema.createRowBatch();
+		orcBatchWriter = new OrcBatchWriter(Arrays.asList(orcSchemaToTableSchema(schema).getTypes()));
+	}
+
+	private TableSchema orcSchemaToTableSchema(TypeDescription orcSchema) {
+		List<String> fieldNames = orcSchema.getFieldNames();
+		List<TypeDescription> typeDescriptions = orcSchema.getChildren();
+		List<TypeInformation> typeInformations = new ArrayList<>();
+
+		typeDescriptions.forEach(typeDescription -> {
+			typeInformations.add(schemaToTypeInfo(typeDescription));
+		});
+
+		return new TableSchema(
+			fieldNames.toArray(new String[fieldNames.size()]),
+			typeInformations.toArray(new TypeInformation[typeInformations.size()]));
+	}
+
+	@Override
+	public void write(T element) throws IOException {
+		Boolean isFill = orcBatchWriter.fill(rowBatch, element);
+		if (!isFill) {
```

Comment:
So when rowBatch is full, you basically throw away the record that's supposed to add?

Reply:
Thanks @xndai for review! Here, ```isFill``` return false if batch is full. So, when the rowBatch is full, we add a row batch to the ORC file.

Reply:
With current implementation, when user call write() and rowBatch is full, you flush the batch to ORC file. But where does the input parameter "element" go?

Reply:
Thanks! I will fix this bug! Very nice question!